eventuals/catch.h

```diff
@@ -12,38 +15,113 @@ namespace eventuals {
 ////////////////////////////////////////////////////////////////////////
 
 struct _Catch final {
-  template <typename K_, typename F_>
+  template <typename K_, typename Error_, typename F_>
+  struct Handler final {
+    using Error = Error_;
+
+    Handler(F_ f)
+      : f_(std::move(f)) {}
+
+    // Helper to convert a catch handler to a new 'K'.
+    template <typename K>
+    auto Convert() && {
+      return Handler<K, Error_, F_>{std::move(f_)};
+    }
+
+    template <typename E>
+    void ConstructAndStart(K_&& k, Interrupt* interrupt, E&& e) {
+      adapted_.emplace(Then(std::move(f_)).template k<Error_>(std::move(k)));
+
+      if (interrupt != nullptr) {
+        adapted_->Register(*interrupt);
+      }
+
+      adapted_->Start(std::forward<E>(e));
+    }
+
+    template <typename E>
+    // NOTE: we're _must_ explicitly take 'k' and 'e' by rvalue reference here
+    // because we do _not_ want the compiler to move the 'k' and 'e' when
+    // calling this function in the event that we _don't_ end up handling the
+    // error.
+    bool Handle(K_&& k, Interrupt* interrupt, E&& e) {
+      // When 'Error_' is 'std::exception_ptr' it indicates we're the
+      // 'all' handler which catches everything and makes a
+      // 'std::exception_ptr' if we don't already have one.
+      if constexpr (std::disjunction_v<
+                        std::is_same<Error_, std::exception_ptr>,
+                        std::is_same<Error_, E>,
+                        std::is_base_of<Error_, E>>) {
+        if constexpr (!std::is_same_v<Error_, std::exception_ptr>) {
+          ConstructAndStart(std::move(k), interrupt, std::forward<E>(e));
+        } else {
+          ConstructAndStart(
+              std::move(k),
+              interrupt,
+              make_exception_ptr_or_forward(std::forward<E>(e)));
+        }
+
+        return true;
```

Comment:
I'd still prefer to explicitly call out the first case related to the comment above, otherwise the comment gets lost and it's hard to discern how the "all" handler works:
```suggestion
      // When 'Error_' is 'std::exception_ptr' it indicates we're the
      // 'all' handler which catches everything but we need to pass a
      // 'std::exception_ptr' in case we don't already have one.
      if constexpr (std::is_same_v<Error_, std::exception_ptr>) {
        ConstructAndStart(
            std::move(k),
            interrupt,
            make_exception_ptr_or_forward(std::forward<E>(e)));
        
        return true;
      } else if constexpr (std::disjunction_v<
                        std::is_same<Error_, E>,
                        std::is_base_of<Error_, E>>) {
        ConstructAndStart(std::move(k), interrupt, std::forward<E>(e));
        return true;
```